File: MicroPython/ClassicController.py
# Author: Alex Ghandhi
# Last Modified: 11 September 2023

"""
This code reads I2C input from a Wii Classic Controller.

Before data can be read, the device must first be decrypted.
This is accomplished with the 'new' method at the wiibrew wiki: http://wiibrew.org/wiki/Wiimote/Extension_Controllers
    write 0x55 to 0xF0
    write 0x00 to 0xFB

Set device into data format 0x03: http://wiibrew.org/wiki/Wiimote/Extension_Controllers/Classic_Controller
    write 0x03 to 0xFE

Controller Client I2C info:
    I2C bus address: 0x52
"""

from machine import Pin, I2C
from time import sleep_ms, sleep_us


class ClassicController(object):
    # I2C Client Address
    bus_addr: int = 0x52

    ##################
    # DUNDER METHODS #
    ##################

    def __init__(self, id: int, sda: int, scl: int, freq: int) -> None:
        """ Initialize ClassicController Object.

        Args
            id:   i2c ID
            sda:  SDA pin
            scl:  SCL pin
            freq: frequency for I2C device
        """
        # The buffer stores the controller output
        self.buffer = bytearray(8)

        # Create I2C client
        self.i2c = I2C(id=id, sda=Pin(sda), scl=Pin(scl), freq=freq)
        sleep_ms(10)

        # Run a bus scan, this should show '[82]' (0x52)
        print(f"Bus Scan Results: {self.i2c.scan()}")
        sleep_ms(10)

        # Decrypt Controller, check the Extension ID and set Data Mode 3
        self.decrypt_controller()
        extensionID = bytearray(6)
        self.i2c.writeto(self.bus_addr, b'\xFA')
        sleep_ms(10)
        self.i2c.readfrom_into(self.bus_addr, extensionID)
        sleep_ms(10)
        print(f"Extension ID: {extensionID}")
        print("Setting Data Mode 3")
        self.set_data_mode_3()

        # Set initial values for calibration variables
        self.joy_LX_center = 0
        self.joy_LY_center = 0
        self.joy_RX_center = 0
        self.joy_RY_center = 0
        self.trigger_L_initial = 0
        self.trigger_R_initial = 0

        # Calibrate Joysticks and Triggers
        print("Calibrating...", end="")
        self.calibrate()
        print("Complete!")

    def __str__(self):
        self.update()
        print("\033c", end='')
        print(f"Left Joy: {self.joy_LX()}, {self.joy_LY()}")
        print(f"Right Joy: {self.joy_RX()}, {self.joy_RY()}")
        print(f"Triggers: L {self.trigger_L()}, R {self.trigger_R()}")
        if (self.button_A() is 1): print("A ", end="")
        if (self.button_B() is 1): print("B ", end="")
        if (self.button_X() is 1): print("X ", end="")
        if (self.button_Y() is 1): print("Y ", end="")
        if (self.button_UP() is 1): print("up ", end="")
        if (self.button_DOWN() is 1): print("down ", end="")
        if (self.button_LEFT() is 1): print("left ", end="")
        if (self.button_RIGHT() is 1): print("right ", end="")
        if (self.button_ZL() is 1): print("ZL ", end="")
        if (self.button_ZR() is 1): print("ZR ", end="")
        if (self.button_L() is 1): print("L ", end="")
        if (self.button_R() is 1): print("R ", end="")
        if (self.button_START() is 1): print("start ", end="")
        if (self.button_SELECT() is 1): print("select ", end="")
        if (self.button_HOME() is 1): print("home ", end="")

    #####################
    # UTILITY FUNCTIONS #
    #####################

    def decrypt_controller(self) -> None:
        self.i2c.writeto(self.bus_addr, b'\xF0\x55')
        sleep_ms(10)
        self.i2c.writeto(self.bus_addr, b'\xFB\x00')
        sleep_ms(10)

    def set_data_mode_3(self) -> None:
        self.i2c.writeto(self.bus_addr, b'\xFE\x03')
        sleep_ms(10)

    def update(self) -> None:
        # Read data from controller and update the output buffer
        # From wiibrew.org: Data is reported in 8 bytes when decrypted
        self.i2c.writeto(self.bus_addr, b'\x00\x00')
        sleep_us(100)
        self.i2c.readfrom_into(self.bus_addr, self.buffer)
        sleep_us(100)

    def calibrate(self) -> None:
        # Find center values for joysticks and triggers
        self.update()
        self.joy_LX_center = self.joy_LX()
        self.joy_LY_center = self.joy_LY()
        self.joy_RX_center = self.joy_RX()
        self.joy_RY_center = self.joy_RY()
        self.trigger_L_initial = self.trigger_L()
        self.trigger_R_initial = self.trigger_R()


    ##################
    # BUTTON GETTERS #
    ##################

    def button_A(self) -> int:
        # Get value of button A
        return int(~(self.buffer[7] >> 4) & 1)

    def button_B(self) -> int:
        # Get value of button B
        return int(~(self.buffer[7] >> 6) & 1)

    def button_X(self) -> int:
        # Get value of button X
        return int(~(self.buffer[7] >> 3) & 1)

    def button_Y(self) -> int:
        # Get value of button Y
        return int(~(self.buffer[7] >> 5) & 1)

    def button_HOME(self) -> int:
        # Get value of button HOME
        return int(~(self.buffer[6] >> 3) & 1)

    def button_START(self) -> int:
        # Get value of button START
        return int(~(self.buffer[6] >> 2) & 1)

    def button_SELECT(self) -> int:
        # Get value of button SELECT
        return int(~(self.buffer[6] >> 4) & 1)

    def button_UP(self) -> int:
        # Get value of button DPAD UP
        return int(~self.buffer[7] & 1)

    def button_DOWN(self) -> int:
        # Get value of button DPAD DOWN
        return int(~(self.buffer[6] >> 6) & 1)

    def button_LEFT(self) -> int:
        # Get value of button DPAD LEFT
        return int(~(self.buffer[7] >> 1) & 1)

    def button_RIGHT(self) -> int:
        # Get value of button DPAD RIGHT
        return int(~(self.buffer[6] >> 7) & 1)

    def button_ZL(self) -> int:
        # Get value of button ZL
        return int(~(self.buffer[7] >> 7) & 1)

    def button_ZR(self) -> int:
        # Get value of button ZR
        return int(~(self.buffer[7] >> 2) & 1)

    def button_L(self) -> int:
        # Get value of button L (bottom-out for trigger)
        return int(~(self.buffer[6] >> 5) & 1)

    def button_R(self) -> int:
        # Get value of button R (bottom-out for trigger)
        return int(~(self.buffer[6] >> 1) & 1)

    def trigger_L(self) -> int:
        # Get value of Left trigger
        return int(self.buffer[4]) if self.buffer[4] > self.trigger_L_initial else 0

    def trigger_R(self) -> int:
        # Get value of Right trigger
        return int(self.buffer[5]) if self.buffer[5] > self.trigger_R_initial else 0

    def joy_LX(self) -> int:
        # Get value of Left joystick x-axis
        return int(self.buffer[0]) - self.joy_LX_center

    def joy_LY(self) -> int:
        # Get value of Left joystick y-axis
        return int(self.buffer[2]) - self.joy_LY_center

    def joy_RX(self) -> int:
        # Get value of Right joystick x-axis
        return int(self.buffer[1]) - self.joy_RX_center

    def joy_RY(self) -> int:
        # Get value of Right joystick y-axis
        return int(self.buffer[3]) - self.joy_RY_center


# Debug: create instance and print controller output
if __name__ == '__main__':
    controller = ClassicController(id=1, sda=6, scl=7, freq=100000)
    while (True):
        print(controller)
        sleep_us(100)

File: C/ClassicController.c
// Author: Alex Ghandhi
// Last Modified: 24 September 2023

/*
This code reads I2C input from a Wii Classic Controller.

Before data can be read, the device must first be decrypted.
This is accomplished with the 'new' method at the wiibrew wiki: http://wiibrew.org/wiki/Wiimote/Extension_Controllers
    write 0x55 to 0xF0
    write 0x00 to 0xFB

Set device into data format 0x03: http://wiibrew.org/wiki/Wiimote/Extension_Controllers/Classic_Controller
    write 0x03 to 0xFE

Controller Client I2C info:
    I2C bus address: 0x52
*/


#include <stdio.h>
#include "ClassicController.h"
#include "pico/stdlib.h"
#include "hardware/i2c.h"


///////////////
// CONSTANTS //
///////////////

// Device I2C Addresses
static const uint I2C_BUS_ADDR = 0x52;
static const uint8_t REGISTER_READ_ADDR = 0x00;

// Controller Decryption I2C Addresses and Write Values
static const uint8_t INIT_MSG_0[] = {0xF0, 0x55};
static const uint8_t INIT_MSG_1[] = {0xFB, 0x00};

// Set Device to Data Mode 3
static uint8_t DATA_MODE_MSG[] = {0xFE, 0x03};

//////////////////////
// GLOBAL VARIABLES //
//////////////////////

// Calibrated values for joysticks and triggers
static uint8_t LX_center, LY_center, RX_center, RY_center, LT_init, RT_init;


//////////////////////////////
// FUNCTION IMPLEMENTATIONS //
//////////////////////////////

// Reads data from the controller and updates the button variables to reflect that data
void ClassicController_update(ClassicController controller) {
    // Array to store output
    uint8_t controller_out[8];

    // Read 8 bytes of data and return an array of those bytes
    i2c_write_blocking(controller.i2c, I2C_BUS_ADDR, &REGISTER_READ_ADDR, 1, false);
    sleep_us(200);
    if (i2c_read_blocking(controller.i2c, I2C_BUS_ADDR, controller_out, 8, false) != 8) {
        printf("\nFailed to read button data from controller\n");
    }

    // Update button values to reflect data from controller_out
    controller.LX          = controller_out[0];
    controller.RX          = controller_out[1];
    controller.LY          = controller_out[2];
    controller.RY          = controller_out[3];
    controller.LT_ANALOG   = controller_out[4];
    controller.RT_ANALOG   = controller_out[5];
    controller.RIGHT       = (~(controller_out[6] >> 7) & 1);
    controller.DOWN        = (~(controller_out[6] >> 6) & 1);
    controller.LT          = (~(controller_out[6] >> 5) & 1);
    controller.SELECT      = (~(controller_out[6] >> 4) & 1);
    controller.HOME        = (~(controller_out[6] >> 3) & 1);
    controller.START       = (~(controller_out[6] >> 2) & 1);
    controller.RT          = (~(controller_out[6] >> 1) & 1);
    controller.ZL          = (~(controller_out[7] >> 7) & 1);
    controller.B           = (~(controller_out[7] >> 6) & 1);
    controller.Y           = (~(controller_out[7] >> 5) & 1);
    controller.A           = (~(controller_out[7] >> 4) & 1);
    controller.X           = (~(controller_out[7] >> 3) & 1);
    controller.ZR          = (~(controller_out[7] >> 2) & 1);
    controller.LEFT        = (~(controller_out[7] >> 1) & 1);
    controller.UP          = (~controller_out[7] & 1);
}

// Calibrates the joysticks and analog triggers
void ClassicController_calibrate(ClassicController controller) {
    // Obtain controller status
    ClassicController_update(controller);

    // Update calibration variables
    LX_center = controller.LX;
    LY_center = controller.LY;
    RX_center = controller.RX;
    RY_center = controller.RY;
    LT_init   = controller.LT;
    RT_init   = controller.RT;
}

// Initialize the controller
ClassicController ClassicController_init(i2c_inst_t *i2c, uint sda, uint scl, uint baudrate) {
    // Initialize I2C functionality
    i2c_init(i2c, baudrate);
    gpio_set_function(sda, GPIO_FUNC_I2C);
    gpio_set_function(scl, GPIO_FUNC_I2C);
    gpio_pull_up(sda);
    gpio_pull_up(scl);

    // Decrypt Controller
    sleep_ms(1);
    i2c_write_blocking(i2c, I2C_BUS_ADDR, INIT_MSG_0, 2, false);
    sleep_ms(1);
    i2c_write_blocking(i2c, I2C_BUS_ADDR, INIT_MSG_1, 2, false);
    sleep_ms(1);

    // Set controller to data report mode 0x03
    i2c_write_blocking(i2c, I2C_BUS_ADDR, DATA_MODE_MSG, 2, false);
    sleep_ms(1);

    // Create a Classic_Controller instance
    ClassicController controller;
    controller.i2c = i2c;

    // Calibrate the joysticks and triggers
    ClassicController_calibrate(controller);
    return controller;
}

// Print currently pressed buttons
void ClassicController_button_report(ClassicController controller) {
    printf("\e[1;1H\e[2J");  // Clear the console window
    printf("Button Report:\n");
    // Report Joystick Values
    printf("Left Joy:\tX: %d\tY: %d\nRight Joy:\tX: %d\tY: %d\n",
            controller.LX,
            controller.LY,
            controller.RX,
            controller.RY
    );
    // Report Trigger Values
    printf("Left Trigger: %d\nRight Trigger: %d\n",
            controller.LT_ANALOG,
            controller.RT_ANALOG
    );
    // Report all other buttons and trigger bottom-out values
    printf("Other Buttons:\n");
    if(controller.UP) printf("UP ");
    if(controller.DOWN) printf("DOWN ");
    if(controller.LEFT) printf("LEFT ");
    if(controller.RIGHT) printf("RIGHT ");
    if(controller.A) printf("A ");
    if(controller.B) printf("B ");
    if(controller.X) printf("X ");
    if(controller.Y) printf("Y ");
    if(controller.START) printf("START ");
    if(controller.SELECT) printf("SELECT ");
    if(controller.HOME) printf("HOME ");
    if(controller.ZL) printf("ZL ");
    if(controller.ZR) printf("ZR ");
    if(controller.LT) printf("LT ");
    if(controller.RT) printf("RT ");
}
